plugins/thunderbolt/fu-thunderbolt-image.c

```diff
@@ -582,11 +693,13 @@ fu_plugin_thunderbolt_validate_image (GBytes  *controller_fw,
 
 	locations = is_host ?
 			    get_host_locations (hw_info->id) :
-			    get_device_locations (hw_info->id);
+			    get_device_locations (hw_info->id, &controller, error);
 	if (locations == NULL) {
-		g_set_error_literal (error,
-				     FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED,
-				     "FW locations to check not found for this controller");
+		/* We may already have an error set so use that then */
+		if (*error == NULL)
+			g_set_error_literal (error,
+					     FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED,
+					     "FW locations to check not found for this controller");
```

Comment:
This section flow for error handling doesn't seem right either.  `get_host_locations` doesn't use an error but `get_device_locations` does.  I'd say you should either split this into an if block around `is_host` or pass in error to `get_host_locations` and then set appropriate error message in `get_host_locations` and `get_device_locations` both.

Then you don't have a scenario of "might have an error" set.  If locations is `NULL` there is certainly an error set and this returns `VALIDATION_FAILED`.